crates/ra_lsp_server/src/main_loop/handlers.rs

```diff
@@ -467,26 +465,43 @@ pub fn handle_rename(world: ServerWorld, params: RenameParams) -> Result<Option<
         .into());
     }
 
-    let renames = world
+    let change = world
         .analysis()
         .rename(FilePosition { file_id, offset }, &*params.new_name)?;
-    if renames.is_empty() {
+    if change.is_none() {
         return Ok(None);
     }
 
-    let mut changes = HashMap::new();
-    for edit in renames {
-        changes
-            .entry(file_id.try_conv_with(&world)?)
-            .or_insert_with(Vec::new)
-            .extend(edit.edit.conv_with(&line_index));
-    }
+    let mut source_change = change.unwrap();
+    let text_document_edits = source_change
+        .source_file_edits
+        .drain(..)
+        .into_iter()
+        .map(|e| e.try_conv_with(&world))
+        .collect::<Result<Vec<TextDocumentEdit>>>();
 
-    Ok(Some(WorkspaceEdit {
-        changes: Some(changes),
+    let text_document_ops = source_change
+        .file_system_edits
+        .drain(..)
+        .into_iter()
+        .map(|e| e.try_conv_with(&world))
+        .collect::<Result<Vec<ResourceOp>>>();
 
-        // TODO: return this instead if client/server support it. See #144
-        document_changes: None,
+    let mut document_changes = Vec::new();
+    document_changes.extend(
+        text_document_edits?
+            .into_iter()
+            .map(DocumentChangeOperation::Edit),
+    );
+    document_changes.extend(
+        text_document_ops?
+            .into_iter()
+            .map(DocumentChangeOperation::Op),
+    );
+
```

Comment:
@matklad 
what's a good way to use `try_conv_with`?

https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_lsp_server/src/conv.rs#L258

Reply:
I think just `source_change.try_conv_with(&server_world)` should work? [This impl](https://github.com/rust-analyzer/rust-analyzer/blob/454cc313589fb17de92d6f3dbf576a5ea5f4adf2/crates/ra_lsp_server/src/conv.rs#L255-L257) should do the trick.

Reply:
`source_change.try_conv_with(&server_world)` returns a `req::SourceChange`. It looks strange to me to get a `req::SourceChange`.

Reply:
Oh, you are right. I think it's fine though: you can extract `WorkspaceEdit` from `SourceChange`.

Reply:
It's not easy. `WorkspaceEdit` doesn't have a label while `SourceChange` do.

Reply:
It's not easy. `WorkspaceEdit` doesn't have a label while `SourceChange` do.

Reply:
we should return `req::SourceChange` just like `handle_join_lines` or `handle_on_enter` instead of `WorkspaceEdit`. am i right?

Reply:
No, we should return `WorkspaceEdit` here. The primary difference between `WorkspaceEndit` and `SourceEdit` is that the letter allows us to position the cursor after applying the edit. We do this by writing custom code on the editor's side. For rename though, we don't need to position cursor, so we can just use the built-in `WorkspaceEdit`.